sdk/core/core/src/az_http_request_builder.c

```diff
@@ -3,48 +3,271 @@
+  }
+
+  if (!az_const_span_is_valid(name) || !az_const_span_is_valid(value) || name.size == 0
+      || value.size == 0) {
+    return AZ_ERROR_ARG;
+  }
+
+  for (size_t i = 0; i < name.size; ++i) {
+    if (name.begin[i] == '\0') {
+      return AZ_ERROR_ARG;
+    }
+  }
+
+  for (size_t i = 0; i < value.size; ++i) {
+    if (value.begin[i] == '\0') {
+      return AZ_ERROR_ARG;
+    }
+  }
+
+  uint8_t * const buffer_end = p_hrb->buffer.begin + p_hrb->buffer.size;
+  uint8_t * new_header_start = buffer_end;
+  {
+    size_t const separator_sizes = sizeof((uint8_t)'\0') + (2 * sizeof((uint8_t)'\0'));
+
+    uint8_t * const headers_start = p_hrb->url.begin + p_hrb->max_url_size;
+
+    size_t const headers_space = buffer_end - headers_start;
+    if (headers_start >= buffer_end || headers_space < separator_sizes) {
+      return AZ_ERROR_BUFFER_OVERFLOW;
+    }
+
+    if (p_hrb->headers_end == 0) {
+      new_header_start = headers_start;
+    } else {
+      uint16_t nheader = 1;
+      for (size_t i = 0; i < headers_space - separator_sizes; ++i) {
+        if (headers_start[i] == '\0' && headers_start[i + 1] == '\0') {
+          assert(i < headers_space);
+          assert(i + 1 < headers_space);
+          assert(i + 2 < headers_space);
+
+          if (nheader >= p_hrb->headers_end) {
+            new_header_start = headers_start + (i + sizeof((uint8_t)'\0')) + 1;
+            break;
+          }
+
+          ++nheader;
+        }
+      }
+    }
+
+    size_t const available_space = buffer_end - new_header_start;
+
+    size_t const name_and_value_size = name.size + value.size;
+    size_t const required_space = name_and_value_size + separator_sizes;
+
+    if (name_and_value_size < name.size || name_and_value_size < value.size
+        || required_space < name_and_value_size || required_space > available_space) {
+      return AZ_ERROR_BUFFER_OVERFLOW;
+    }
+
+    az_const_span const new_header_span = { .begin = new_header_start, .size = required_space };
+    if (az_const_span_is_overlap(new_header_span, name)
+        || az_const_span_is_overlap(new_header_span, value)) {
+      return AZ_ERROR_ARG;
+    }
+  }
+
+  assert(new_header_start != buffer_end);
+
+  memcpy(new_header_start, name.begin, name.size);
+  new_header_start += name.size;
+
+  *new_header_start = '\0';
+  ++new_header_start;
+
+  memcpy(new_header_start, value.begin, value.size);
+  new_header_start += value.size;
+
+  new_header_start[0] = '\0';
+  new_header_start[1] = '\0';
+
+  ++(p_hrb->headers_end);
+
   return AZ_OK;
 }
 
 az_result az_http_request_builder_mark_retry_headers_start(az_http_request_builder * const p_hrb) {
-  (void)p_hrb;
+  p_hrb->retry_headers_start = p_hrb->headers_end;
   return AZ_OK;
 }
 
 az_result az_http_request_builder_remove_retry_headers(az_http_request_builder * const p_hrb) {
-  (void)p_hrb;
+  p_hrb->headers_end = p_hrb->retry_headers_start;
   return AZ_OK;
 }
+
+az_result az_http_request_builder_get_header(
```

Comment:
I think this function is at the end not so good...  
Function basically traverse all headers one by one to get the header at position `index`.  Operation is O(n).  However, if this function is used by some consumer (like the curl adapter) to loop all headers, the time complexity becomes O(n2) since getting each header will cost O(n). 

I think it would be good to keep the size of each header in control block, or together with the header in the buffer, the we can introduce an utility function to receive the pointer to a header, reads its size and returns the position of the next header by doing a simple offset instead of traversing from the start every time.

Reply:
It's called `iterator` when you keep current position and have a function `get_next`. :-)

Reply:
All this function should do is return the az_span at the specified index; I'd expect this to also be about 2-3 lines of code.

Reply:
Done.